numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -4610,20 +4509,77 @@ set_typeinfo(PyObject *dict)
      *          CFLOAT, CDOUBLE, CLONGDOUBLE,
      *          OBJECT, STRING, UNICODE, VOID,
      *          DATETIME,TIMEDELTA#
+     * #NAME = Bool,
+     *         Byte, UByte, Short, UShort, Int, UInt,
+     *         Long, ULong, LongLong, ULongLong,
+     *         Half, Float, Double, LongDouble,
+     *         CFloat, CDouble, CLongDouble,
+     *         Object, String, Unicode, Void,
+     *         Datetime, Timedelta#
+     * #kind = GENBOOL,
+     *         SIGNED, UNSIGNED, SIGNED, UNSIGNED, SIGNED, UNSIGNED,
+     *         SIGNED, UNSIGNED, SIGNED, UNSIGNED,
+     *         FLOATING, FLOATING, FLOATING, FLOATING,
+     *         COMPLEX, COMPLEX, COMPLEX,
+     *         OBJECT, STRING, UNICODE, VOID,
+     *         DATETIME, TIMEDELTA#
+     * #endian = |*3, =*15, |, |, =, |, =*2#
+     * #flags= 0*18, NPY_OBJECT_DTYPE_FLAGS,0, NPY_NEEDS_INIT, 0*3#
+     * #fromtype = npy_bool,
+     *             npy_byte, npy_ubyte, npy_short, npy_ushort, npy_int, npy_uint,
+     *             npy_long, npy_ulong, npy_longlong, npy_ulonglong,
+     *             npy_half, npy_float, npy_double, npy_longdouble,
+     *             npy_cfloat, npy_cdouble, npy_clongdouble,
+     *             PyObject *, char, npy_ucs4, char,
+     *             npy_datetime, npy_timedelta#
+     * #DO_ELSIZE = 1*19, 0*3, 1*2#
      */
 
-    /**begin repeat1
-     *
-     * #name2 = HALF, DATETIME, TIMEDELTA#
-     */
+#if USE_DTYPE_AS_PYOBJECT
+    dtype = PyObject_New(PyArray_Descr, &PyArrayDescr_Type);
+#else
+    dtype = PyObject_GC_New(PyArray_Descr, &PyArrayDescr_Type);
```

Comment:
Somewhat random: I'm trying to follow along (and failing), but looking up this function, the documentation [1] notes that this is for functions that have `Py_TPFLAGS_HAVE_GC` set -- but you don't have that and also do not do `PyObject_GC_Track`. If it is not needed yet, I would not use the `GC_new` version.

[1] https://docs.python.org/3.5/c-api/gcsupport.html#c.PyObject_GC_New

Reply:
Thanks, well spotted.

The default `tp_free` for type objects is `PyObject_GC_Del`. I set the one for `PyArray_DescrType` to `PyObject_Free` and changed the `PyObject_GC_New` allocations to PyObject_New`.